api/src/main/java/org/pmiops/workbench/monitoring/MeasurementBundle.java

```diff
@@ -0,0 +1,131 @@
+package org.pmiops.workbench.monitoring;
+
+import com.google.common.collect.ImmutableMap;
+import io.opencensus.metrics.data.AttachmentValue;
+import io.opencensus.metrics.data.AttachmentValue.AttachmentValueString;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.pmiops.workbench.monitoring.attachments.AttachmentKey;
+import org.pmiops.workbench.monitoring.views.OpenCensusView;
+
+/**
+ * Simple bundle class to avoid dealing with lists of pairs of maps. This situation can occur in
+ * classes that have multiple kinds of signals with different attribute maps among subsets. In order
+ * to avoid excessive proliferation of (confusing) argument list types, we try to consolidate here.
+ */
+public class MeasurementBundle {
+  private final Map<OpenCensusView, Number> monitoringViews;
+  private final Map<AttachmentKey, String> attachmentKeyToString;
+
+  private MeasurementBundle(
+      Map<OpenCensusView, Number> monitoringViews,
+      Map<AttachmentKey, String> attachmentKeyToString) {
+    this.monitoringViews = monitoringViews;
+    this.attachmentKeyToString = attachmentKeyToString;
+  }
+
+  public Map<OpenCensusView, Number> getMeasurements() {
+    return monitoringViews;
+  }
+
+  public Map<String, AttachmentValue> getAttachments() {
+    return fromAttachmentKeyToString(attachmentKeyToString);
+  }
+
+  // In our system, it's more convenient to work with the AttachmentKey
+  // enum directly and String values. For some reason we have to wrap strings
+  // in AttachmentValueStrings now (though naked Strings are still supported but
+  // deprecated).
+  private static Map<String, AttachmentValue> fromAttachmentKeyToString(
+      Map<AttachmentKey, String> attachmentKeyStringMap) {
+    return attachmentKeyStringMap.entrySet().stream()
+        .map(
+            e ->
+                new SimpleImmutableEntry<String, AttachmentValue>(
+                    e.getKey().getKeyName(), AttachmentValueString.create(e.getValue())))
+        .collect(ImmutableMap.toImmutableMap(Entry::getKey, SimpleImmutableEntry::getValue));
+  }
+
+  public static Builder builder() {
```

Comment:
If we had Immutables, we'd basically be done with this file now. Probably just define a validate() function for the builder to use.

Reply:
It's really extensible. For instnace with a `check` method: https://immutables.github.io/immutable.html#precondition-check-method

It's really not an either-or situation, and as long as we have any Java it still makes sense to do this. Yes it's way too heavy to bring in Kotlin if this is all you wanted.